### PR TITLE
Refactoring test to use latest Parser

### PR DESCRIPTION
**Pull request by poojapi**
First take on the refactoring. The number of fails increased with the new parser. Wanted to get reviewed before going further.

### CODE REVIEW

**Comment by kmadathil** [comment]
I couldn't run this from the tests branch. Could you check if you've pushed all your commits? 
Some initial comments - 
  I'd prefer if the actual checking happens in `test_splits` for clarity. 
  Please set the input encoding of Parser to ITRANS explicitly - this test uses ITRANS wheras most others use SLP1

**Comment by poojapi** [comment]
I'll do the changes as indicated asap.

**Comment by poojapi** [comment]
@kmadathil Can you tell me what error you got while running? I think I found a bug in which the parser errors out with input_encoding given as 'ITRANS'.  I will check based on your error message.

**Comment by kmadathil** [comment]
@poojapi - When I ran tests from that branch last week, I saw a 100% error
(exception) result with zero % pass or fail

**Comment by kmadathil** [comment]
I think #136 must be the cause of that, @poojapi . 
Temporarily, please comment out the lines that I've called out in #136 in api.py
```
if self.input_encoding is not None:
            self.input_encoding = SCHEMES[self.input_encoding]
        if self.output_encoding is not None:
            self.output_encoding = SCHEMES[self.output_encoding]
```
and please use the following in your test.

```
from sanskrit_parser.base.sanskrit_base import ITRANS
# No quotes around ITRANS
lexan = Parser(input_encoding=ITRANS, replace_ending_visarga='s')
```

**Comment by poojapi** [comment]
@kmadathil I had to use input_encoding as ITRANS and output_encoding as SLP1 to make it work. Please review again.

**Comment by kmadathil** [comment]
Good, this is working now. Let's wait for Avinash to look at #136 before merging, though.
In the meantime, please check the fail output to see if you find anything evidently suspicious. Is our script extracting phrases from the input itx file wrongly somehow?  Can you try a few fails at random on the command line and see if the command line output fails to produce the desired output too (if it doesn't, there's something wrong with the test script).
One last thing I'd try is to add score=False to the Parser instantiation and see if that reduces or increases the number of fails.

**Comment by kmadathil** [comment]
Just noticed this: `replace_ending_visarga=None`  is the right way to call the Parser for sandhi splitting. In fact, since None is the default value, and I can't see any reason why we'd call this api with any other value than None for sandhi or vakya analysis, it's best not even use it in the test.

**Comment by poojapi** [comment]
As indicated, I removed the replace_ending_visarga. The results are as follows ->

Without scoring=false
Pass = 342, Split Pass = 0, Fail = 1056, Skip = 0, Error = 0

With scoring=false
Pass = 301, Split Pass = 0, Fail = 1097, Skip = 0, Error = 0

I tested with same random 10-15 failures. All of them are true failures. I'll push my latest code changes.

**Comment by kmadathil** [comment]
Good! I'll merge as soon as we resolve #136. 
Would you like to move on to the other two tests?

**Comment by kmadathil** [comment]
Another possible direction is to look into the BG test script itself. The state machine code I wrote is pretty naive, and was meant to be a quick-and-dirty start. It clearly gets a whole bunch of things wrong. The structure of the itx file is verse followed by sandhi split followed by anvaya. Sometimes the anvaya isn't there (for things line sa~njaya uvAca). 
Further, I haven't done some basic things like removing spaces to make the input string more regular. Maybe there's something to explore there. Not all of our rules work with spaces, which is known - #97 
So, whichever direction you prefer works. @poojapi

**Comment by avinashvarna** [comment]
I looked at the first "fail" and the only thing that is different is that "aDyAyaH" is split as "'aDi', 'AyaH''", which is technically correct. This is probably related to the scoring as well, since it seems to weight this split more than retaining "aDyAyaH" as a single word. 
At least on the command line, increasing the number of splits to 25 produces the "aDyAyaH" split near the middle of the pack, but changing the max_splits argument in test_splits doesn't seem to make this case "pass". May need to take a closer look at the test_splits function.

**Comment by avinashvarna** [comment]
At least for the first one that I am checking, the "fail" with max_splits = 25 is because the expected split contains a '.' and an empty string;
```
DEBUG:root:Expected ['atha', 'prathamaH', 'adhyAyaH', '.', '', 'arjuna', 'viShAda', 'yogaH']
```

The expected splits may need additional cleanup.

File: tests/generate_bg_Lexical_Analyzer_pass_fail.py [hunk 1]
```diff
@@ -146,24 +152,60 @@ def _in_splits(s, splits):
...
+                    splitKhandaClean = splitKhanda.strip("'")
+                    cleanedSplits.append(splitKhandaClean)
+                
+                # check if the cleaned split is present in slpSplits.
+                # if not found, set didNotFind to True which is used later for sandhi
+                # failure check
+                split = cleanedSplits
+                for slpSplit in slpSplits:
+                    if (slpSplit not in split):
+                        didNotFind = True
```
Comment (avinashvarna) [inline tests/generate_bg_Lexical_Analyzer_pass_fail.py:190]: This loop may miss the case where a word in the split occurs twice, but is present only once in the generated splits.

**Comment by poojapi** [comment]
@kmadathil I'll work on improving the current test, complete it and then pick the others.
@avinashvarna I'll make the changes as indicated. Also, will adopt a different method for the missing loop case you reviewed. 

Thanks for the review! Will update the PR soon.

**Comment by avinashvarna** [comment]
Thanks @poojapi . I pushed a fix for #136 to this PR, so that it would help you with refactoring the tests as well. I hope this doesn't interfere with your changes.

**Comment by kmadathil** [comment]
Thanks @poojapi . Please look into the design issues in the state machine code. That would've resulted in the issue that @avinashvarna pointed out. 

Something that I noticed - in this snippet in the SM code, the `prev = line` is incorrect (shouldn't be there) 
```
                        continue
                    if line[-2:] == " .":  # Found | , non verse
                        prev = line
                        pnum = lnum
                        state = "S0"
                        r = [prev[:-2], line.split(" ")[:-1], prev, line, False,
                             basename, pnum]
```

**Comment by kmadathil** [comment]
Closing as obsolete.

**Pull request closed without merging**